Add support for F3DEX in gbi.h, add support for all GBI's in gfxdis.

glankk opened the pull request.

glankk committed:
add support for f3dex, revert gfxdis to f3dex2-only for now

valandil commented:
Thanks glank! I'll review and test the changes later tonight, this weekend at the latest.

glankk commented on include/n64/gbi.h:
```diff
@@ -1,8 +1,14 @@
  * n64 graphics microcode interface library
- * compatible with f3dex2 and s2dex2
- * -glank
+ * compatible with fast3d, f3dex, f3dex2, s2dex, and s2dex2
```
For compatibility with SGI's`gbi.h`, use Fast3D by default and select F3DEX and F3DEX2 with `F3DEX_GBI` and `F3DEX_GBI_2` respectively.

glankk commented on include/n64/gbi.h:
```diff
@@ -11,66 +17,74 @@
+#endif
+
+/* commands for fast3d and f3dex */
+#if defined(F3D_GBI) || defined(F3DEX_GBI)
```
Microcode-specific definitions are enclosed in a consistently styled `#if defined(...` block. No `#ifndef`, `#elif` or `#else` are used, nor are there any nested conditional blocks; this makes the code easier to understand, as it is immediately clear which microcode(s) each block apply to.

valandil replied:
Good, I like this. I forgot about this syntax of the C preprocessor.

glankk commented on include/n64/gbi.h:
```diff
@@ -143,34 +176,39 @@
-#define G_CLIPPING                    (gI_(0b0)<<0)
 
-#else
+/* geometry mode for fast3d */
```
Common definitions appear first, followed by conditional blocks of microcode-specific definitions. Separating definitions by GBI compatibility makes it clear which definitions need special consideration when dealing with specific microcodes.

glankk commented on include/n64/gbi.h:
```diff
@@ -906,49 +926,9 @@
 #define G_MWO_bLIGHT_1                gI_(0x04)
-
-#ifdef F3D_GBI
-#define G_MWO_bLIGHT_1                gI_(0x04)
```
Removed multiple definition of `G_MWO_bLIGHT_1`.

glankk commented on include/n64/gbi.h:
```diff
@@ -1791,77 +1818,7 @@ gsDPSetTile(fmt,siz,line,tmem,      \
                                           gF_(uls,12,12)|gF_(ult,12,0),       \
                                           gF_(tile,3,24)|gF_(lrs,12,12)|      \
                                           gF_(lrt,12,0))
-#define gsSP1Triangle(v0,v1,v2,flag)  gO_(G_TRI1,                             \
```
`gsSP1Triangle` is incompatible between Fast3D and F3DEX2. Added a Fast3D-specific definition (line 2021).

glankk commented on include/n64/gbi.h:
```diff
@@ -1870,124 +1827,17 @@ gsSPBranchLessZrg(branchdl,vtx,     \
-
-// The Fast3D ucode does not support gSPLoadGeometryMode.
-#ifndef F3D_GBI
-#define gsSPLoadGeometryMode(mode)    gsSPGeometryMode(~gI_(0),mode)
```
`gSPLoadGeometryMode(...)` can be emulated with a `gSPClearGeometryMode(~0)` followed by a `gSPSetGeometryMode(...)`. Added because why not (line 2051).

valandil replied:
Good idea. This will also remove some conditional preprocessing that I added to some parts of gfx.c.

glankk commented on include/n64/gbi.h:
```diff
@@ -1870,124 +1827,17 @@ gsSPBranchLessZrg(branchdl,vtx,     \
-#endif
-
 #define gsSPLine3D(v0,v1,flag)        gsSPLineW3D(v0,v1,0,flag)
-#define gsSPLineW3D(v0,v1,wd,flag)    gO_(G_LINE3D,                           \
```
`gsSPLineW3D` is incompatible between Fast3D and F3DEX2. Added a Fast3D-specific definition (line 2027).

glankk commented on include/n64/gbi.h:
```diff
@@ -11,66 +17,74 @@
-
-#define G_SPNOOP                      0x00
-#define G_MTX                         0x01
-#define G_RESERVED0                   0x02
```
The `G_RESERVED*` codes do not seem to have any meaning at all (as opposed to the `G_SPECIAL_*` codes which do have some meaning, though I don't what they are). Maybe they should be included for completeness, but I've chosen to remove them.

glankk commented on include/n64/gbi.h:
```diff
@@ -820,77 +858,59 @@
 #define G_MWO_SEGMENT_0               gI_(0x00)
-
-#ifdef F3D_GBI
-#define G_MWO_SEGMENT_1               gI_(0x01)
```
This `gbi.h` defines `G_MWO_SEGMENT_*` as their true offsets for consistency with the meaning of `G_MWO_*` definitions. There is no GBI incompatibility between Fast3D and F3DEX2 for these definitions, so I've removed this microcode-specific block.

glankk commented on include/n64/gbi.h:
```diff
@@ -1870,124 +1827,17 @@ gsSPBranchLessZrg(branchdl,vtx,     \
-
-#ifdef F3D_GBI
-
-#define gsSPPopMatrix(param)          gO_(G_POPMTX,0x1,param)
```
Another rogue `0x1`? Changed to `0` (line 2057).

glankk commented on include/n64/gbi.h:
```diff
@@ -1870,124 +1827,17 @@ gsSPBranchLessZrg(branchdl,vtx,     \
-                                          gF_(16,16,0),                       \
-                                          (char *)(mtr)+48)
-
-#define gsSPSetGeometryMode(mode)     gO_(G_SETGEOMETRYMODE,  0x1, gI_(mode))
```
I'm not sure where these `0x1`'s come from, did you mean to say `0x0`? Changed to `0` (line 2049 and 2050).

valandil replied:
Upon review, I have no idea where those `0x1` come from either. They were there from my very first commit...

glankk commented on include/n64/gbi.h:
```diff
@@ -2035,35 +1885,13 @@ gsSPBranchLessZrg(branchdl,vtx,     \
                                                G_MWO_a##Lightnum,packedcolor),\
                                       gsMoveWd(G_MW_LIGHTCOL,                 \
                                                G_MWO_b##Lightnum,packedcolor)
-#define gsSPTexture(sc,tc,level,    \
```
`gsSPTexture` is incompatible between Fast3D and F3DEX2. Added a Fast3D-specific definition (line 2060).

glankk commented on include/n64/gbi.h:
```diff
@@ -1870,124 +1827,17 @@ gsSPBranchLessZrg(branchdl,vtx,     \
-                                          mptr),                              \
-                                      gO_(G_MOVEMEM,gF_(G_MV_MATRIX_2,8,16)|  \
-                                          gF_(16,16,0),                       \
-                                          (char *)(mtr)+16),                  \
```
Changed `mtr` typo to `mptr` (line 2044, 2046, and 2048).

glankk commented on include/n64/gbi.h:
```diff
@@ -2170,33 +1982,9 @@ gsSPScisTextureRectangleFlip(ulx,   \
-
-#ifdef F3D_GBI
-#define gsDPHalf1(wordhi)             gO_(G_RDPHALF_2,0,wordhi)
-#define gsDPHalf2(wordlo)             gO_(G_RDPHALF_CONT,0,wordlo)
```
Where does this change come from? As far as I can tell, these definitions are identical in Fast3D and F3DEX2. Reverted.

glankk commented on include/n64/gbi.h:
```diff
@@ -2238,37 +2004,6 @@ gsSPSetOtherMode(opc,shift,length,  \
-                                          gF_(offset,16,0),data)
-#endif
-
-#define gsMoveMem(size,index,       \
```
`gsMoveMem` is incompatible between Fast3D and F3DEX2. Added a Fast3D-specific definition (line 2072).

glankk reviewed (commented):
I added some comments describing some of the changes and the reasoning behind them.

valandil commented:
In my testing, it seems that this works fine on the 2.0h version of the microcode (i.e. on the Shindou edition), but the textures either do not appear or are corrupted on the 2.0d version of the microcode (SM64 J and U). 

When I first tested with my PR, it was the other way around.

I'll try to confirm this later tonight, and show video of the issue.

valandil commented:
[Here's](https://youtu.be/U165awi61S4) the footage of issue.The first two tests are with your PR, while the second test is with mine.

glankk commented:
That's odd. Looking through various manuals and docs, I haven't been able to find any mentions of GBI incompatibilities between revisions of the Fast3D microcode. I'd like to do some testing to find the precise command(s) that is causing this, but unfortunately I'm not able to do that at the moment. If you have the time, I'd appreciate your help with this.

valandil commented:
Yeah, no problem. 

I'll try to isolate the commands used in the hack that changed in this PR.

glankk committed:
add support for early versions of fast3d and f3dex, fix fast3d 1tri and line3d

glankk committed:
gfxdis: f3d and f3dex support, split into subprogram for each gbi, bump to 0.2

glankk commented:
Turns out early versions of Fast3D (and apparently F3DEX too) have a different definition of `gsSPPerspNormalize` and `gsDPWord` (i.e. `gsDPHalf1` and `gsDPHalf2`, as defined in your original Fast3D commit). I found that adding two new GBI's (something like `F3D_BETA_GBI` and `F3DEX_BETA_GBI`) with only minor changes was a bit clumsy, so these are now selected by specifying `-DF3D_BETA` in addition to `-DF3D_GBI` or `-DF3DEX_GBI`. I've also added all supported GBI's to gfxdis and split it into one subprogram for each GBI (`gfxdis.f3d`, `gfxdis.f3db`, `gfxdis.f3dex`, `gfxdis.f3dexb`, `gfxdis.f3dex2` respectively).

`gbi.h` should hopefully be compatible with both SM64-J/U (`-DF3D_GBI -DF3D_BETA`) and the Shindou edition (`-DF3D_GBI`) now.

joeydumont commented:
Thanks for all the work glank! I'll test this tonight.

joeydumont commented:
glank, you are a gentleman and a scholar! All 3 versions, U, J, and S now display textures properly. 

Thank you for taking my contributions and elevating them to that level of quality, it's been quite educative for me. I'll make sure that my next PR is better than my first two.

glankk commented:
Great! Thanks for testing. Merging.

glankk committed:
Merge branch 'master' into feature-ucodes
